### PR TITLE
Add call coins

### LINKED ISSUE #2154: Add coins parameter to call ABI
## Use case

I want to create a smart contract that allows buying some new token using Massa coins. To achieve this, my smart contract exports a buy_token function: 

`buy_token(amount)`

Problem: how can the SC check that I sent it enough Massa coins for it to attribute a certain amount of tokens to me in its database ? We could use `transfer_coins` beforehand to give the SC enough coins, but the SC won't be able to to tell those coins came from me and were directed towards buying tokens.

## Solution

Add a `coins` parameter to the `call(...)` ABI:

`call(target_addr, target_function, parameters, coins)`

When called, the ABI will start by spending `coins` massa from the sender address' SCE ledger balance and crediting them to the target address, defining a context parameter `call_coins` that lets the target SC know how many coins were transferred from the caller during that specific call using a `get_call_coins() -> u64` ABI.

### PR DESCRIPTION
**Pull request by gterzian**
Close #2154

### CODE REVIEW

**Comment by gterzian** [comment]
@damip Can you please take a first look? The changes to the cargo file will be removed once https://github.com/massalabs/massa-sc-runtime/pull/82 merges...

File: massa-execution/Cargo.toml [hunk 1]
```diff
@@ -17,7 +17,7 @@ serde_json = "1.0"
 thiserror = "1.0"
 tokio = { version = "1.11", features = ["full"] }
 tracing = {version = "0.1", features = ["max_level_debug", "release_max_level_debug"] }
-assembly_simulator = { git = "https://github.com/massalabs/massa-sc-runtime", tag = "v0.2.8" }
+assembly_simulator = {path = "../../sc-runtime"}
```
Comment (damip) [inline massa-execution/Cargo.toml:20]: ok this is temporary, let's not forget to put it back

File: massa-execution/src/interface_impl.rs [hunk 2]
```diff
@@ -127,7 +131,7 @@ impl Interface for InterfaceImpl {
     fn set_data_for(
         &self,
         address: &assembly_simulator::Address,
-        key: &str,
+        key: &assembly_simulator::MassaHash,
```
Comment (damip) [inline massa-execution/src/interface_impl.rs:134]: Not your fault but https://github.com/massalabs/massa/pull/2145 is still not merged, and those changes are gonna create conflicts. For eample, `key` is not pre-hashed anymore.

Could you base your work on https://github.com/massalabs/massa/pull/2145 ?

File: massa-execution/src/interface_impl.rs [hunk 3]
```diff
@@ -145,7 +149,7 @@ impl Interface for InterfaceImpl {
         }
     }
 
-    fn get_data(&self, key: &str) -> Result<Bytecode> {
+    fn get_data(&self, key: &assembly_simulator::MassaHash) -> Result<Bytecode> {
```
Comment (damip) [inline massa-execution/src/interface_impl.rs:152]: see comment above

File: massa-execution/src/interface_impl.rs [hunk 4]
```diff
@@ -158,7 +162,7 @@ impl Interface for InterfaceImpl {
         }
     }
 
-    fn set_data(&self, key: &str, value: &Bytecode) -> Result<()> {
+    fn set_data(&self, key: &assembly_simulator::MassaHash, value: &Bytecode) -> Result<()> {
```
Comment (damip) [inline massa-execution/src/interface_impl.rs:165]: see comment above

**Review by damip — DISMISSED** [review]
A couple more things:
* make sure that context.call_coins is reset to 0 in clear_and_update_context()
* make sure that context.call_coins is reset to 0 in prepare_context()

File: massa-execution/src/interface_impl.rs (hunk 2 above)
Comment (gterzian) [reply massa-execution/src/interface_impl.rs:134]: I think either PR can rebase and fix conflicts, depending on which one merges first.

**Comment by gterzian** [comment]
> 

    make sure that context.call_coins is reset to 0 in clear_and_update_context()
    make sure that context.call_coins is reset to 0 in prepare_context()

`clear_and_update_context` seems enough right? @damip

File: massa-execution/src/interface_impl.rs (hunk 2 above)
Comment (damip) [reply massa-execution/src/interface_impl.rs]: https://github.com/massalabs/massa/pull/2145 is merged. You should rebase this one on main

**Comment by damip** [comment]
> > 
> 
> ```
> make sure that context.call_coins is reset to 0 in clear_and_update_context()
> make sure that context.call_coins is reset to 0 in prepare_context()
> ```
> 
> `clear_and_update_context` seems enough right? @damip

not really: when there are multiple successive ExecuteSC in a block (at the same slot), it's important to reset it for each one

**Comment by gterzian** [comment]
> not really: when there are multiple successive ExecuteSC in a block (at the same slot), it's important to reset it for each one

Ok, thanks for the info, done.

**Review by yvan-sraka — DISMISSED** [review]

**Comment by damip** [comment]
bors merge

**Comment by damip** [comment]
bors cancel

**Review by adrien-zinger — APPROVED** [review]

**Comment by adrien-zinger** [comment]
bors merge

**Pull request merged**